### PR TITLE
Migrate remaining tests to async/await syntax

### LINKED ISSUE #569: Migrate tests to async/await syntax
As this is a relatively big change, the plan is to do it incrementally.

The following test could be used for some inspiration: https://github.com/hazelcast/hazelcast-nodejs-client/blob/9c1c2d10b8d5a2dc74fa21ddca3676a3e66cfb25/test/ClientRedoEnabledTest.js

### PR DESCRIPTION
**Pull request by sudiptog81**
Migrated remaining tests in the Hazelcast Node.js Client to use the async/await syntax. Also performed cleanup of named exports in the relevant tests. Let me know if there are any suggestions or additional migrations.

Closes #569.

### CODE REVIEW

File: test/serialization/config/FactoriesTest.js [hunk 1]
```diff
@@ -57,27 +55,17 @@ describe('FactoriesTest', function () {
...
-                });
-            });
+    it('should be configured programmatically', async function () {
+        client = await Client.newHazelcastClient(createConfig(cluster.id));
+        const map = await client.getMap('furkan');
+        await map.put('foo', new Foo('elma'));
+        await map.put('address', new Address('Sahibiata', 42000, 'Konya', 'Turkey'));
+        const res = await map.get('foo');
+        expect(res.foo).to.be.equal('elma');
+        const res_1 = await map.get('address');
```
Comment (mdumandag) [inline test/serialization/config/FactoriesTest.js:65]: I guess we can define `res` as let and reuse it

Comment (sudiptog81) [reply test/serialization/config/FactoriesTest.js:65]: Addressing this in the next commit.

File: test/nearcache/NearCacheSimpleInvalidationTest.js [hunk 2]
```diff
@@ -32,67 +32,55 @@ describe('NearCacheSimpleInvalidationTest', function () {
...
+                expect(stats.missCount).to.equal(entryCount);
+                expect(stats.entryCount).to.equal(entryCount);
+                let putPromise = Promise.resolve();
+                for (let i = 0; i < entryCount; i++) {
+                    putPromise = putPromise.then(map.put.bind(map, '' + i, 'changedvalue', undefined));
+                }
+                await putPromise;
+                getPromise = Promise.resolve();
+                for (let i = 0; i < entryCount; i++) {
+                    getPromise = getPromise.then(map.get.bind(map, '' + i));
```
[the remark this answers is not in the document]
Comment (sudiptog81) [reply test/nearcache/NearCacheSimpleInvalidationTest.js:77]: Addressing this in the next commit.

File: test/nearcache/NearCacheSimpleInvalidationTest.js [hunk 3]
```diff
@@ -32,67 +32,55 @@ describe('NearCacheSimpleInvalidationTest', function () {
...
+                    await getPromise;
+                    await map.get('' + i);
+                }
+                await getPromise;
+                let stats = map.nearCache.getStatistics();
+                expect(stats.missCount).to.equal(entryCount);
+                expect(stats.entryCount).to.equal(entryCount);
+                let putPromise = Promise.resolve();
+                for (let i = 0; i < entryCount; i++) {
+                    putPromise = putPromise.then(map.put.bind(map, '' + i, 'changedvalue', undefined));
```
[the remark this answers is not in the document]
Comment (sudiptog81) [reply test/nearcache/NearCacheSimpleInvalidationTest.js:72]: Addressing this in the next commit.

File: test/nearcache/NearCacheSimpleInvalidationTest.js [hunk 4]
```diff
@@ -32,67 +32,55 @@ describe('NearCacheSimpleInvalidationTest', function () {
...
-                    return getPromise;
-                }).then(function () {
-                    const stats = map.nearCache.getStatistics();
-                    expect(stats.entryCount).to.equal(entryCount);
-                    expect(stats.hitCount).to.equal(0);
-                    expect(stats.missCount).to.equal(entryCount * 2);
-                });
+                let map = await client.getMap(mapName);
+                let getPromise = Promise.resolve();
+                for (let i = 0; i < entryCount; i++) {
```
[the remark this answers is not in the document]
Comment (sudiptog81) [reply test/nearcache/NearCacheSimpleInvalidationTest.js:62]: Addressing this in the next commit.

File: test/nearcache/LostInvalidationsTest.js [hunk 5]
```diff
@@ -76,101 +73,73 @@ describe('LostInvalidationTest', function () {
...
-                const promise = (function (key) {
-                    return map.get(key).then((val) => {
-                        return expect(val).to.equal(key + entryCount);
-                    });
-                })(i);
-                promises.push(promise);
-            }
-            return Promise.all(promises);
-        });
+        let map = await client.getMap(mapName);
```
[the remark this answers is not in the document]
Comment (sudiptog81) [reply test/nearcache/LostInvalidationsTest.js:111]: Addressing this in the next commit.

File: test/nearcache/LostInvalidationsTest.js [hunk 6]
```diff
@@ -76,101 +73,73 @@ describe('LostInvalidationTest', function () {
...
-            return Util.promiseWaitMilliseconds(1000);
-        }).then(function () {
-            return map.get(key);
-        }).then(function (result) {
-            return expect(result).to.equal(updatedval);
-        });
+        let map = await client.getMap(mapName);
+        await Util.promiseWaitMilliseconds(100);
+        const invalidationHandlerStub = blockInvalidationEvents(client, map, 1);
+        map = await modifyingClient.getMap(mapName);
```
[the remark this answers is not in the document]
Comment (sudiptog81) [reply test/nearcache/LostInvalidationsTest.js:97]: Addressing this in the next commit.

File: test/integration/DistributedObjectsTest.js [hunk 7]
```diff
@@ -32,149 +31,72 @@ describe('DistributedObjectsTest', function () {
...
-            });
+    it('get distributed objects should clear local instances of destroyed proxies', async function () {
+        const otherClient = await Client.newHazelcastClient({ clusterName: cluster.id });
+        const map = await client.getMap(Util.randomString());
+        const set = await otherClient.getSet(Util.randomString());
+        const queue = await client.getQueue(Util.randomString());
+        const objects = await client.getDistributedObjects();
+        expect(toNamespace(objects)).to.have.deep.members(toNamespace([map, set, queue]));
+        await map.destroy();
+        const objects_1 = await client.getDistributedObjects();
```
[the remark this answers is not in the document]
Comment (sudiptog81) [reply test/integration/DistributedObjectsTest.js:90]: Addressing this in the next commit.

File: test/heartbeat/HeartbeatFromClientTest.js [hunk 8]
```diff
@@ -52,43 +50,31 @@ describe('HeartbeatFromClientTest', function () {
...
-            }, 1000);
-            return Util.promiseLater(15000, () => {});
-        }).then(function () {
-            clearInterval(pushTask);
-            expect(connectionClosedEventCount).to.equal(0);
-            return client1.shutdown();
-        }).then(function () {
-            return client2.shutdown();
+        await RC.startMember(cluster.id);
+        client1 = await Client.newHazelcastClient(clientConfig);
```
[the remark this answers is not in the document]
Comment (sudiptog81) [reply test/heartbeat/HeartbeatFromClientTest.js:54]: Addressing this in the next commit.

File: test/flakeid/FlakeIdGeneratorProxyTest.js [hunk 9]
```diff
@@ -30,126 +30,94 @@ describe("FlakeIdGeneratorProxyTest", function () {
...
-                    expect(idList[i - 1].equals(idList[i]), 'Expected ' + idList[i - 1] + ' ' + idList[i] + 'to be different.')
-                        .to.be.false;
-                }
-            });
+    it('newId returns a unique long', async function () {
+        flakeIdGenerator = await client.getFlakeIdGenerator('test');
+        let promise = Promise.resolve();
+        const idList = [];
+        for (let i = 0; i < 10; i++) {
+            await promise;
```
[the remark this answers is not in the document]
Comment (sudiptog81) [reply test/flakeid/FlakeIdGeneratorProxyTest.js:72]: Addressing this in the next commit.

File: test/nearcache/LostInvalidationsTest.js [hunk 10]
```diff
@@ -76,101 +73,71 @@ describe('LostInvalidationTest', function () {
...
+        entries = [];
+        for (let i = 0; i < entryCount; i++) {
+            entries.push([i, i + entryCount]);
+        }
+        await mp.putAll(entries);
+        unblockInvalidationEvents(client, invalidationHandlerStub);
+        await Util.promiseWaitMilliseconds(2000);
+        const promises = [];
+        for (let i = 0; i < entryCount; i++) {
+            const promise = (function (key) {
```
Comment (mdumandag) [inline test/nearcache/LostInvalidationsTest.js:134]: We can also replace this with `await` syntax

**Review by mdumandag — COMMENTED** [review]
Thanks a lot for the changes @sudiptog81. One extra place in which we can use `await` caught my attention. Could you also update that? Then, we can run the test suite

File: test/serialization/SerializationServiceTest.js [hunk 11]
```diff
@@ -15,9 +15,9 @@
  */
 'use strict';
 
-const expect = require('chai').expect;
-const SerializationService = require('../../lib/serialization/SerializationService').SerializationServiceV1;
-const SerializationConfigImpl = require('../../lib/config/SerializationConfig').SerializationConfigImpl;
+const { expect } = require('chai');
+const { SerializationService } = require('../../lib/serialization/SerializationService');
```
Comment (sudiptog81) [inline test/serialization/SerializationServiceTest.js:19]:
```suggestion
const { SerializationService: SerializationServiceV1 } = require('../../lib/serialization/SerializationService');
```

File: test/ssl/ClientSSLAuthenticationTest.js [hunk 12]
```diff
@@ -16,14 +16,14 @@
...
 const chaiAsPromised = require('chai-as-promised');
 chai.use(chaiAsPromised);
+const expect = chai.expect;
 const fs = require('fs');
 const path = require('path');
 
-const RC = require('./../RC');
-const { Client, IllegalStateError } = require('../../');
+const RC = require('../RC');
+const { Client, IllegalStateError } = require('../../lib');
```
Comment (sudiptog81) [inline test/ssl/ClientSSLAuthenticationTest.js:26]:
```suggestion
const { Client, IllegalStateError } = require('../../');
```

File: test/ssl/ClientSSLAuthenticationTest.js [hunk 13]
```diff
@@ -37,11 +37,9 @@ describe('ClientSSLAuthenticationTest', function () {
         markEnterprise(this);
     });
 
-    function createMemberWithXML(serverXML) {
-        return RC.createCluster(null, fs.readFileSync(serverXML, 'utf8')).then(function (cl) {
-            cluster = cl;
-            return RC.startMember(cluster.id);
-        });
+    async function createMemberWithXML(serverXML) {
+        const cluster = await RC.createCluster(null, fs.readFileSync(serverXML, 'utf8'));
```
Comment (sudiptog81) [inline test/ssl/ClientSSLAuthenticationTest.js:41]:
```suggestion
        cluster = await RC.createCluster(null, fs.readFileSync(serverXML, 'utf8'));
```

**Review by sudiptog81 — COMMENTED** [review]
I might have overlooked some things in the mentioned tests. The following pointers could give an insight:

* The `getClientStatisticsFromServer` method in `StatisticsTest ` should have been `async` as we are executing on the cluster, tests were failing due to improper promise handling. I have fixed that and the tests are now passing. It is interesting how it passed in #628.
* `ReliableTopicTest` passes locally without any changes.

I am trying to address the aforementioned issues in the next commit. However, I might need some guidance for the SSL tests as I am not able to understand what went wrong.

**Comment by mdumandag** [comment]
verify

**Comment by mdumandag** [comment]
The commit should solve the problems related to serialization service and SSL authentication tests.  The other SSL tests failed due to a timeout but this PR does not change its file. Maybe something not properly closed in other tests caused that. I will look more into it. In the meantime, I will restart the test suite

**Review by mdumandag — APPROVED** [review]
Looks good to me after the recent changes. Thanks a lot @sudiptog81. I will re-run the test suit one more time to be sure and then merge

**Comment by mdumandag** [comment]
verify

**Pull request merged**